docs/modules/ROOT/pages/explanation/decisions/usernames.adoc

```diff
@@ -0,0 +1,57 @@
+= `User` object names in the OpenShift cluster
+
+Problem::
+We need to define an identifier which we use to name users which are synchronized from Keycloak to the OpenShift cluster.
++
+The identifier should have the following properties:
++
+* It should be stable for the lifetime of the user in Keycloak.
+* It must be suitable for use as an OpenShift resource name (namely for the OpenShift `User` resource)
+* It must be used to configure OpenShift `Group` membership when synchronizing groups from Keycloak.
+
++
+In the implementation, this identifier is selected by providing an appropriate attribute name for the OpenID claim `preferred_username` in the OpenShift OpenID Connect identity provider configuration.
+
+
+Alternatives::
+* Use the `Username` property of Keycloak user objects.
++
+With this approach we can use the existing https://github.com/redhat-cop/group-sync-operator[redhat-cop/group-sync-operator] as it uses the Keycloak `Username` property to populate the user list in the OpenShift `Group` objects it creates.
+Additionally, `oc whoami` provides users with the `Username` they chose when registering their user.
+However, Users can't change their `Username` after registration, as the OpenShift `User` and `Identity` objects, which are created when the users first logs in, aren't updated when the user changes their `Username` in Keycloak.
+
+* Use the `ID` property of Keycloak user objects.
++
+With this approach, users can freely change their `Username` in Keycloak.
++
+However, their `Username` doesn't appear anywhere in the OpenShift cluster with this approach and the output of `oc whoami` is the Keycloak user's `ID`.
+Additionally, group synchronization from Keycloak to OpenShift will require a custom implementation to create `Group` objects which are populated with the group members' `ID` properties, as the group-sync operator currently only supports using the Keycloak user's `Username` field when creating `Group` objects.
+
+* Use the `Email` property of Keycloak user objects.
++
+With this approach, users can't change their E-Mail address after signing up, for the same reasons as they can't change their `Username` property when using that property (see first alternative).
+
+
+Decision::
+* We use the `Username` property of Keycloak user objects as the name for the resulting OpenShift `User` objects
+* We allow users to freely choose their `Username` when signing up
+* We disallow users to change their `Username` after creation of the user in Keycloak
```

Comment:
I'm not sure if it makes sense to have all of these in a single document, or if we should split them out

Reply:
I kinda agree. The `We use the Username property of Keycloak user objects as the name for the resulting OpenShift User objects` decision makes sense, the other two are separate.